File: src/api/oauth/discord.js
const axios = require('axios');
const qs = require('querystring');

const app = require("../../app");
const logger = require("../../logger");
const user_dao = require("../../dao/user");
const setupuser = require("../../util/setupuser");

const { MANAGEMENT, UPLOAD, OVERWRITE } = require("../../permissions");
const createjwt = require("../../util/createjwt");

app.get('/api/oauth_callback/discord', function(req, res){
  logger.debug("GET /api/oauth_callback/discord", req.query);

  const data = {
    client_id: process.env.DISCORD_APP_ID,
    client_secret: process.env.DISCORD_APP_SECRET,
    code: req.query.code,
		grant_type: 'authorization_code',
		redirect_uri: process.env.BASE_URL + "/api/oauth_callback/discord",
    'scope': 'identify email connections'
  };

  const options = {
    headers: {
      "Accept": "application/json",
			"Content-Type": "application/x-www-form-urlencoded"
    }
  };

	let user_info;
	axios.post("https://discord.com/api/oauth2/token", qs.stringify(data), options)
  .then(r => {
		console.log(r.data);
    if (!r.data.access_token){
      console.log(r);
      throw new Error("No access_token received!");
    }

		return axios.get("https://discord.com/api/users/@me", {
      headers: {
        "Authorization": "Bearer " + r.data.access_token
      }
    })
		.then(r => {
			user_info = r.data;
			console.log(user_info);
			// id, username, email, verified
			return user_dao.get_by_external_id(user_info.id);
		})
		.then(user => {
			if (user){
				if (user.type != "DISCORD") {
					throw new Error("User already exists with another type!");
				}
				// valid user
				return user;
			} else {
				// create new user
				return user_dao.create({
					name: user_info.username,
					type: "DISCORD",
					hash: "", // no local password
					mail: user_info.email,
					external_id: user_info.id
				})
				.then(setupuser);
			}
		});
  })
	.then(user => {
		const token = createjwt(user, [UPLOAD, OVERWRITE, MANAGEMENT]);
		res.redirect(process.env.BASE_URL + "/#/oauth/" + token);
	})
	.catch(e => {
		console.log(e);
		res.send(e.message);
	});

});

File: web/oauth/discord.go
package oauth

import (
	"blockexchange/core"
	"blockexchange/types"
	"bytes"
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
)

type DiscordResponse struct {
	ID       string `json:"id"`
	Username string `json:"username"`
	Email    string `json:"email"`
}

type DiscordOauth struct {
}

func (o *DiscordOauth) RequestAccessToken(code string, cfg *core.Config) (string, error) {
	q := url.Values{}
	q.Add("client_id", cfg.DiscordOAuthConfig.ClientID)
	q.Add("client_secret", cfg.DiscordOAuthConfig.Secret)
	q.Add("redirect_uri", cfg.BaseURL+"/api/oauth_callback/discord")
	q.Add("code", code)
	q.Add("grant_type", "authorization_code")
	q.Add("scope", "identify email connections")
	fmt.Println(q.Encode())

	buf := bytes.NewBufferString(q.Encode())

	req, err := http.NewRequest("POST", "https://discord.com/api/oauth2/token", buf)
	if err != nil {
		return "", err
	}

	req.Header.Set("Accept", "application/json")
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")

	client := http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return "", err
	}

	tokenData := AccessTokenResponse{}
	err = json.NewDecoder(resp.Body).Decode(&tokenData)
	if err != nil {
		return "", err
	}

	return tokenData.AccessToken, nil
}

func (o *DiscordOauth) RequestUserInfo(access_token string, cfg *core.Config) (*OauthUserInfo, error) {
	req, err := http.NewRequest("GET", "https://discord.com/api/users/@me", nil)
	if err != nil {
		return nil, nil
	}

	req.Header.Set("Accept", "application/json")
	req.Header.Set("Authorization", "Bearer "+access_token)

	client := http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}

	userData := DiscordResponse{}
	err = json.NewDecoder(resp.Body).Decode(&userData)
	if err != nil {
		return nil, err
	}

	fmt.Println(userData)
	info := OauthUserInfo{
		Name:       userData.Username,
		Type:       types.UserTypeDiscord,
		Email:      userData.Email,
		ExternalID: userData.ID,
	}

	return &info, nil
}
